sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp

```diff
@@ -485,7 +485,25 @@ __esimd_svm_atomic0(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
     ;
 #else
 {
-  __ESIMD_UNSUPPORTED_ON_HOST;
+  __ESIMD_DNS::vector_type_t<Ty, N> Output = 0;
+
+  for (int AddrIdx = 0; AddrIdx < N; AddrIdx += 1) {
+    if (pred[AddrIdx] == 0) {
+      // Skip Output vector elements correpsonding to
+      // predicates whose value is zero
+      continue;
+    }
+    // Keeping original values for return + 'load'
+    Output[AddrIdx] = *((Ty *)addrs[AddrIdx]);
+    if constexpr (Op == __ESIMD_NS::atomic_op::inc) {
+      __ESIMD_DNS::atomic_add_fetch<Ty>((Ty *)addrs[AddrIdx],
+                                        static_cast<Ty>(1));
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::dec) {
+      __ESIMD_DNS::atomic_sub_fetch<Ty>((Ty *)addrs[AddrIdx],
+                                        static_cast<Ty>(1));
+    }
```

Comment:
This is wrong. The update becomes non-atomic. There can be other updates happening between writing to Output above and the actual atomic operation.
```suggestion
    if constexpr (Op == __ESIMD_NS::atomic_op::inc) {
      Output[AddrIdx] = __ESIMD_DNS::atomic_add_fetch<Ty>((Ty *)addrs[AddrIdx],
                                        static_cast<Ty>(1));
    } else if constexpr (Op == __ESIMD_NS::atomic_op::dec) {
      Output[AddrIdx] = __ESIMD_DNS::atomic_sub_fetch<Ty>((Ty *)addrs[AddrIdx],
                                        static_cast<Ty>(1));
    }
```